Add script to add the RN Swift packages to xcodeproject

cipolleschi opened the pull request:
Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds an higeher level function that can be used to integrate local swift packages into Xcodeproject files.

This will also be helpful for autolinking packages.

## Changelog:
[Internal] -

Differential Revision: D81778453

cortinico reviewed (changes requested):
Review automatically exported from Phabricator review in Meta.

cipolleschi committed:
Add simple function to handle symlinks

Summary:
This change adds a simple function that manages symlinks. This is used through the stack of changes to avoid code repetition.

## Context
SwiftPM is very picky in how the header structure must be.
In order to preserve the import/include statements as much as possible when building from source, we can recreate the header structure in a temporary folder inside the react-native package using symlinks.

In this way, users can still modify the headers and build RNTester and HelloWorld using SwiftPM without breaking changes.

## Changelog:
[Internal] -

Differential Revision: D82202860

Reviewed By: cortinico

cipolleschi committed:
Add function to extract headers from a folder

Summary:
This change add a utility function to extract headers (both .h and .hpp) from a folder. It also allow to exclude some specific folders (e.g.: `tests`, non supported platforms, ...)

## Changelog:
[Internal] -

Differential Revision: D82205692

Commit:
Create script to create hard links for headers from a source path to a destination path (#53620)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53620

## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function that creates links between a source folder and a destination folder.

## Changelog:
[Internal] -

Differential Revision: D81778469

Reviewed By: cortinico

Commit:
Create utility function to create the right folder structure for headers in the ReactApple folder (#53629)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53629

## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding a function to properly create the header structure for the headers in ReactApple that requires a special case.

## Changelog:
[Internal] -

Differential Revision: D81778471

Reviewed By: cortinico

Commit:
add function hard links for headers from ReactCodegen

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function to create hard links to Codegen Headers so that other libraries can access them.

## Changelog:
[Internal] -

Differential Revision: D81778450

Commit:
Create utility function to create the right folder structure for headers in the ReactCommon folder (#53631)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53631

## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function that creates links for the ReactCommon folder that has a complex structure and complex mapping requirements

## Changelog:
[Internal] -

Differential Revision: D81778470

Reviewed By: cortinico

Commit:
Add function to create hard links for headers of the third-party dependencies

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding an helper function that creates links for the ThirdPartyDependencies that React native and libraries might need to access.

## Changelog:
[Internal] -

Differential Revision: D81778459

Commit:
Create the orchestration function to coordinate the creation of the hard links

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we are adding a function that is used to orchestrate the helpers function defined in previous changes.

## Changelog:
[Internal] -

Differential Revision: D81778457

Commit:
Create function to create hardlink to header required to build React Native Core

Summary:
## Context

One of the quirk of SwiftPM is that the packages has to have access to the headers they need. Usually this is solved by properly setting the header_search_path. However, in  SwiftPM, we are not allowed to use headers search path that escape the package itself (basically, header search path can't start with `../`).

To work around this limitation we are recreating the correct Header structure by using hardlinks to the actual headers.

## Changed

In this change we did two things:
1. we move the helper functions to a headers-utils.js file (and the tests to the `headers-utils-test.js` file)
2. we added a function that coordinates the creation of the header links *and added tests for it)

## Changelog:
[Internal] -

Differential Revision: D81778464

Commit:
Add Function to update the PBXFrameworkBuildPhase

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change does two things:
- moves the printFilesForBuildPhase to xcode-core-utils to make it easier to mock the function
- Adds a function to update the package reference section to configure the list of build files for the FrameworkBuildStep

## Changelog:
[Internal] -

Differential Revision: D81778472

Commit:
Add Function to convert pbx file to JSON

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

Apple provides a helpful CLI program called `plutils` that can take a xcodeproj and convert it to json. We use that to parse the xcodeproject.

## Changelog:
[Internal] -

Differential Revision: D81778448

Commit:
Add Function to deintegrate local SwiftPM packages from Xcode project

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

We are adding a function that deintegrate local Swift Packages. This helps to make sure we can have an idempotent script and that we don't add the same package twice.

## Changelog:
[Internal] -

Differential Revision: D81778458

Commit:
Add helper functions to serialize specific objects of the project

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds some helper function that can be used to serialize some simple Xcodeproject objects back to string.
This will be used in future diffs to rewrite the Xcodeproj file.

## Changelog:
[Internal] -

Differential Revision: D81778445

Commit:
Add helper functions to add local packages to xcodeproject

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change does two things:
- moves the function that generate the object id to a different file
- adds a function to add the local Swift packages to the project

Moving the generateObjectId function to a separate file is a required step to mock it in the tests.

## Changelog:
[Internal] -

Differential Revision: D81934363

Commit:
Add Function to update the package reference section

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds a function to update the package reference section so that Xcode can properly load the packages in the project.

## Changelog:
[Internal] -

Differential Revision: D81778463

Commit:
Add function to generate random Xcode ids

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

All the Xcode project elements have an id of 24 hex characters. This change adds a function to generate the id randomly.

## Changelog:
[Internal] -

Differential Revision: D81778473

Commit:
Add Function to add missing sections in case the original project don't have them (#53648)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53648

## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds a function to add missing section to the project. This function is used by a coordination function in a future diff on the stack.

## Changelog:
[Internal] -

Differential Revision: D81778443

Commit:
Add Functions to coordinate the update of the xcode project.

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This is the last function required to manipulate the Xcodeproject file. It is an high-level function that orchestrate all the changes that are needed in the xcodeproj file

## Changelog:
[Internal] -

Differential Revision: D82020721

Commit:
Add Functions to manipulate the update of the xcode project.

Summary:
## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change:
- moves the print functions and the addMissingSections  function to a separate file, to make it easier to mocke them in tests
- add a function that coordinates the manipulation of the xcodeproj

## Changelog:
[Internal] -

Differential Revision: D81778451

cipolleschi committed:
Add script to add the RN Swift packages to xcodeproject (#53668)

Summary:
Pull Request resolved: https://github.com/facebook/react-native/pull/53668

## Context

Apple does not have a tool from CLI to add a package to
a project, therefore we need to manually modify the Xcodeproject to integrate some Swift packages into it.

## Changed

This change adds an higeher level function that can be used to integrate local swift packages into Xcodeproject files.

This will also be helpful for autolinking packages.

## Changelog:
[Internal] -

Differential Revision: D81778453
